tags-config-service-impl/src/main/java/org/hypertrace/tag/config/service/TagConfigServiceImpl.java

```diff
@@ -0,0 +1,97 @@
+import org.hypertrace.tag.config.service.v1.UpdateTagRequest;
+import org.hypertrace.tag.config.service.v1.UpdateTagResponse;
+
+@Slf4j
+public class TagConfigServiceImpl extends TagConfigServiceGrpc.TagConfigServiceImplBase {
+  private final ConfigServiceCoordinator configServiceCoordinator;
+
+  public TagConfigServiceImpl(Channel configChannel) {
+    configServiceCoordinator = new ConfigServiceCoordinatorImpl(configChannel);
+  }
+
+  @Override
+  public void createTag(
+      CreateTagRequest request, StreamObserver<CreateTagResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      CreateTag createTag = request.getTag();
+      Tag tag =
+          Tag.newBuilder().setId(UUID.randomUUID().toString()).setKey(createTag.getKey()).build();
+      Tag createdTag = configServiceCoordinator.upsertTag(requestContext, tag);
+      responseObserver.onNext(CreateTagResponse.newBuilder().setTag(createdTag).build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void getTag(GetTagRequest request, StreamObserver<GetTagResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      String tagId = request.getId();
+      Tag tag = configServiceCoordinator.getTag(requestContext, tagId);
+      responseObserver.onNext(GetTagResponse.newBuilder().setTag(tag).build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void getTags(GetTagsRequest request, StreamObserver<GetTagsResponse> responseObserver) {
+    RequestContext requestContext = RequestContext.CURRENT.get();
+    List<Tag> tagList = configServiceCoordinator.getAllTags(requestContext);
+    responseObserver.onNext(GetTagsResponse.newBuilder().addAllTags(tagList).build());
+    responseObserver.onCompleted();
+  }
+
+  @Override
+  public void updateTag(
+      UpdateTagRequest request, StreamObserver<UpdateTagResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      Tag updatedTagInReq = request.getTag();
+      configServiceCoordinator.getTag(requestContext, updatedTagInReq.getId());
+      Tag updatedTagInRes = configServiceCoordinator.upsertTag(requestContext, request.getTag());
+      responseObserver.onNext(UpdateTagResponse.newBuilder().setTag(updatedTagInRes).build());
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(e);
+    }
+  }
+
+  @Override
+  public void deleteTag(
+      DeleteTagRequest request, StreamObserver<DeleteTagResponse> responseObserver) {
+    try {
+      RequestContext requestContext = RequestContext.CURRENT.get();
+      String tagId = request.getId();
+      configServiceCoordinator.getTag(requestContext, tagId);
```

Comment:
I am curious about this one. Do we truly need to check if tag exists before delete it? This operations don't seem to happen transactionally so after you get a tag and before you delete it, tag might be gone already. If delete is idempotent (which I believe it should), then I would just attempt to delete.

Reply:
@kamaleshnneerasa mentioned that there was a NPE when you attempt to delete a non-existent tag-id. You are right that we should fix it appropriately in root config service such that all config services are benefitted by this change.